clang/lib/DPCT/CubRule.cpp

```diff
@@ -0,0 +1,1354 @@
+                                         bool FuncCallUsed) {
+  std::string Repl;
+  auto DC = CE->getDirectCallee();
+  std::string FuncName = DC->getNameAsString();
+  if (FuncName == "ThreadLoad") {
+    auto FuncArgs = CE->getArgs();
+    const Expr *InData = FuncArgs[0];
+    ExprAnalysis InEA(InData);
+    Repl = "*(" + InEA.getReplacedString() + ")";
+    emplaceTransformation(new ReplaceStmt(CE, Repl));
+  } else if (FuncName == "ThreadStore") {
+    auto FuncArgs = CE->getArgs();
+    const Expr *OutputIterator = FuncArgs[0];
+    const Expr *Value = FuncArgs[1];
+    ExprAnalysis ItrEA(OutputIterator);
+    ExprAnalysis ValueEA(Value);
+    Repl =
+        "*(" + ItrEA.getReplacedString() + ") = " + ValueEA.getReplacedString();
+    emplaceTransformation(new ReplaceStmt(CE, Repl));
+  }
+}
+
+void CubRule::processWarpLevelFuncCall(const CallExpr *CE, bool FuncCallUsed) {
+  std::string Repl;
+  size_t WarpSize = 32;
+  auto DC = CE->getDirectCallee();
+  std::string FuncName = DC->getNameAsString();
+  if (FuncName == "ShuffleIndex") {
+    auto TA = DC->getTemplateSpecializationArgs();
+    if (!TA)
+      return;
+    WarpSize = TA->get(0).getAsIntegral().getExtValue();
+    std::string ValueType =
+        TA->get(1).getAsType().getUnqualifiedType().getAsString();
+    auto MemberMask = CE->getArg(2);
+    auto Mask = dyn_cast<IntegerLiteral>(MemberMask);
+    if (Mask && Mask->getValue().getZExtValue() == 0xffffffff) {
+      const Expr *Value = CE->getArg(0);
+      const Expr *Lane = CE->getArg(1);
+      ExprAnalysis ValueEA(Value);
+      ExprAnalysis LaneEA(Lane);
+      auto DeviceFuncDecl = getImmediateOuterFuncDecl(CE);
+      Repl = DpctGlobalInfo::getSubGroup(CE, DeviceFuncDecl) + ".shuffle(" +
+             ValueEA.getReplacedString() + ", " + LaneEA.getReplacedString() +
+             ")";
+      emplaceTransformation(new ReplaceStmt(CE, Repl));
+      if (DeviceFuncDecl) {
+        auto DI = DeviceFunctionDecl::LinkRedecls(DeviceFuncDecl);
+        if (DI) {
+          DI->addSubGroupSizeRequest(WarpSize, CE->getBeginLoc(), "shuffle");
+        }
+      }
+    } else {
+      report(CE->getBeginLoc(), Diagnostics::API_NOT_MIGRATED, false,
+             "cub::" + FuncName);
+    }
+  }
+}
+
+void CubRule::processCubFuncCall(const CallExpr *CE, bool FuncCallUsed) {
+  const auto *DC = CE->getDirectCallee();
+  if (!DC)
+    return;
+
+  llvm::StringRef FuncName = DC->getName();
+
+  if (FuncName == "ShuffleIndex") {
+    processWarpLevelFuncCall(CE, FuncCallUsed);
+  } else if (FuncName == "ThreadLoad" || FuncName == "ThreadStore") {
+    processThreadLevelFuncCall(CE, FuncCallUsed);
+  } else if (isCubDeviceFuncName(FuncName)) {
+    processDeviceLevelFuncCall(CE, FuncCallUsed);
+  }
+}
+
+void CubRule::processBlockLevelMemberCall(const CXXMemberCallExpr *BlockMC) {
+  if (!BlockMC || !BlockMC->getMethodDecl()) {
+    return;
+  }
+  std::string Repl, NewFuncName, ParamList, InitRepl, OpRepl, Indent,
```

Comment:
For functinality of this file,  I see some function is moved from [clang/lib/DPCT/ASTTraversal.cpp], so please ask old author "wang, hao3" to have a review.

Reply:
+1